Lesson15/hw_lesson15.py

```diff
@@ -0,0 +1,207 @@
+        return sqr
+
+    def remove_positives(self, some_list: list):
+        r_positives = list(filter(lambda num: num < 0, some_list))
+        return r_positives
+
+    def filter_leaps(self, some_list: list):
+        f_leaps = list(filter(lambda num: num % 4 == 0, some_list))
+        return f_leaps
+
+
+m = Mathematician()
+
+assert m.square_nums([7, 11, 5, 4]) == [49, 121, 25, 16]
+
+assert m.remove_positives([26, -11, -8, 13, -90]) == [-11, -8, -90]
+
+assert m.filter_leaps([2001, 1884, 1995, 2003, 2020]) == [1884, 2020]
+
+
+"""
+Task 3
+Product Store
+
+Write a class Product that has three attributes:
+
+type
+name
+price
+
+Then create a class ProductStore, which will have some Products and will operate with all products in the store.
+All methods, in case they can’t perform its action, should raise ValueError with appropriate error information.
+
+Tips: Use aggregation/composition concepts while implementing the ProductStore class.
+You can also implement additional classes to operate on a certain type of product, etc.
+
+Also, the ProductStore class must have the following methods:
+
+add(product, amount) - adds a specified quantity of a single product
+with a predefined price premium for your store(30 percent)
+set_discount(identifier, percent, identifier_type=’name’) - adds a discount for all products specified
+by input identifiers (type or name). The discount must be specified in percentage
+sell_product(product_name, amount) - removes a particular amount of products from the store if available,
+in other case raises an error. It also increments income if the sell_product method succeeds.
+get_income() - returns amount of many earned by ProductStore instance.
+get_all_products() - returns information about all available products in the store.
+get_product_info(product_name) - returns a tuple with product name and amount of items in the store.
+"""
+
+
+class Product:
+
+    c = 0
+
+    def __init__(self, types: str, name: str, price: float):
+        self.types = types
+        self.name = name
+        self.price = price
+        self.__class__.c += 1
+
+    def __str__(self):
+        return self.types + ', ' + self.name + ', ' + str(self.price)
+
+
+class ProductStore:
+
+    def __init__(self):
+        self.goods = dict()
+
+    def add(self, product, amount):
+        self.goods[product.name] = [product.types, product.name, round(product.price * 1.3, 2), amount]
+        return tuple(self.goods[product.name])
+
+    def set_discount(self, identifier, percent):
+        a = self.goods[identifier]
+        a.append(f'{percent} %')
+        self.goods.update({self.goods.get(self, identifier): a})
+        return self.goods
+
+    def sell(self, product_name, amount):
+        a = self.goods[product_name]
+        a[3] = a[3] - amount
+        self.goods.update({self.goods.get(self, product_name): a})
+        return (product_name, self.goods[product_name])
+
+    @staticmethod
+    def get_income(cls):
```

Comment:
We just should make a property with `money_earned_total` and increase it on selling the product. There is a typo in the task description... ;)

Reply:
Changed.